consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/statemachine/QbftBlockHeightManager.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Responsible for starting/clearing Consensus rounds at a given block height. One of these is
+ * created when a new block is imported to the chain. It immediately then creates a Round-0 object,
+ * and sends a Proposal message. If the round times out prior to importing a block, this class is
+ * responsible for creating a RoundChange message and transmitting it.
+ */
+public class QbftBlockHeightManager implements BaseQbftBlockHeightManager {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final IbftRoundFactory roundFactory;
+  private final RoundChangeManager roundChangeManager;
+  private final BlockHeader parentHeader;
+  private final BlockTimer blockTimer;
+  private final QbftMessageTransmitter transmitter;
+  private final MessageFactory messageFactory;
+  private final Map<Integer, RoundState> futureRoundStateBuffer = Maps.newHashMap();
+  private final FutureRoundProposalMessageValidator futureRoundProposalMessageValidator;
+  private final Clock clock;
+  private final Function<ConsensusRoundIdentifier, RoundState> roundStateCreator;
+  private final BftFinalState finalState;
+
+  private Optional<PreparedCertificate> latestPreparedRoundArtifacts = Optional.empty();
+
+  private QbftRound currentRound;
+
+  public QbftBlockHeightManager(
+      final BlockHeader parentHeader,
+      final BftFinalState finalState,
+      final RoundChangeManager roundChangeManager,
+      final IbftRoundFactory ibftRoundFactory,
+      final Clock clock,
+      final MessageValidatorFactory messageValidatorFactory,
+      final MessageFactory messageFactory) {
+    this.parentHeader = parentHeader;
+    this.roundFactory = ibftRoundFactory;
+    this.blockTimer = finalState.getBlockTimer();
+    this.transmitter =
+        new QbftMessageTransmitter(messageFactory, finalState.getValidatorMulticaster());
+    this.messageFactory = messageFactory;
+    this.clock = clock;
+    this.roundChangeManager = roundChangeManager;
+    this.finalState = finalState;
+
+    futureRoundProposalMessageValidator =
+        messageValidatorFactory.createFutureRoundProposalMessageValidator(
+            getChainHeight(), parentHeader);
+
+    roundStateCreator =
+        (roundIdentifier) ->
+            new RoundState(
+                roundIdentifier,
+                finalState.getQuorum(),
+                messageValidatorFactory.createMessageValidator(roundIdentifier, parentHeader));
+
+    currentRound = roundFactory.createNewRound(parentHeader, 0);
+    if (finalState.isLocalNodeProposerForRound(currentRound.getRoundIdentifier())) {
+      blockTimer.startTimer(currentRound.getRoundIdentifier(), parentHeader);
+    }
+  }
+
+  @Override
+  public void handleBlockTimerExpiry(final ConsensusRoundIdentifier roundIdentifier) {
+    if (roundIdentifier.equals(currentRound.getRoundIdentifier())) {
+      currentRound.createAndSendProposalMessage(clock.millis() / 1000);
```

Comment:
nit: divide by `1000L` as millis returns a long value.

Reply:
line 236 is more readable way to convert to seconds I think.

Reply:
done